compiler/rustc_lint/src/types.rs

```diff
@@ -486,6 +508,58 @@ fn lint_literal<'tcx>(
     }
 }
 
+fn lint_nan<'tcx>(
+    cx: &LateContext<'tcx>,
+    e: &'tcx hir::Expr<'tcx>,
+    binop: hir::BinOp,
+    l: &'tcx hir::Expr<'tcx>,
+    r: &'tcx hir::Expr<'tcx>,
+) {
+    fn is_nan(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+        let expr = expr.peel_blocks().peel_borrows();
+        match expr.kind {
+            ExprKind::Path(qpath) => {
+                let Some(def_id) = cx.typeck_results().qpath_res(&qpath, expr.hir_id).opt_def_id() else { return false; };
+
+                matches!(cx.tcx.get_diagnostic_name(def_id), Some(sym::f32_nan | sym::f64_nan))
+            }
+            _ => false,
+        }
+    }
+
+    if is_nan(cx, l) {
+        cx.emit_spanned_lint(
+            INVALID_NAN_COMPARISONS,
+            e.span,
+            match binop.node {
+                hir::BinOpKind::Eq | hir::BinOpKind::Ne => InvalidNanComparisons::EqNe {
+                    suggestion: InvalidNanComparisonsSuggestion {
+                        nan_plus_binop: l.span.until(r.span),
```

Comment:
What if `l.span` and `r.span` come from different expansions? Should we use the ancestors inside `e.span`?

Reply:
> What if `l.span` and `r.span` come from different expansions? Should we use the ancestors inside `e.span`?

It doesn't work. :-(

To fix the issue I've conditioned the suggestion on the suggest-ability of the spans and if we cannot use then we just display a help text. Doing something more felt unnecessary, the help is IMO sufficient.